File: simulator/teams.py
from enum import Enum
import copy


class Teams:

    class Color(Enum):
        BLUE = 1
        RED = 2

        def __str__(self):
            match self:
                case self.BLUE:
                    return "\x1b[38;5;39m"
                case self.RED:
                    return "\x1b[38;5;196m"
                case _:
                    return ""

    def __init__(self):
        self.team_book = {}  # maps team_color -> [list of combatants]
        self.reverse_team_book = {}  # maps combatant -> team_color
        self.id_to_combatant = {-1: None}  # maps combatant id -> combatant
        self.team_color_codes = {}

    def add_combatant_to_team(self, combatant, team_color):
        try:
            self.id_to_combatant[combatant.id] = combatant
            self.reverse_team_book[combatant] = team_color
            self.team_book[team_color].append(combatant)
        except KeyError:
            self.team_book[team_color] = [combatant]
        combatant.add_team(team_color)

    def replace_combatant(self, combatant_old, combatant_new):
        """
        Helper function for wildshape
        """
        self.reverse_team_book[combatant_new] = self.reverse_team_book[combatant_old]
        del self.reverse_team_book[combatant_old]
        self.team_book[self.reverse_team_book[combatant_new]].remove(combatant_old)
        self.team_book[self.reverse_team_book[combatant_new]].append(combatant_new)
        del self.id_to_combatant[combatant_old.id]
        self.id_to_combatant[combatant_new.id] = combatant_new

    def get_team_color_code(self, combatant):
        return str(self.reverse_team_book[combatant])

    def get_team_color(self, combatant):
        return self.reverse_team_book[combatant]

    def get_surviving_teams(self):
        return [name for name, combatants in self.team_book.items() if any(ch.is_alive() for ch in combatants)]

    def get_death_count(self):
        return tuple(len(self.team_book[color]) - sum(ch.is_alive() for ch in self.team_book[color]) for color in
                 [self.Color.BLUE, self.Color.RED])

    def get_team_colors(self):
        return self.team_book.keys()

    def are_allies(self, first, second):
        return True if self.reverse_team_book[first] == self.reverse_team_book[second] else False

    def are_enemies(self, first, second):
        return False if self.reverse_team_book[first] == self.reverse_team_book[second] else True

    def get_team(self, combatant):
        return self.reverse_team_book[combatant]

    def get_allies(self, combatant):
        team_members = copy.copy(self.team_book[self.reverse_team_book[combatant]])
        team_members.remove(combatant)  # Remove self
        return team_members

    def get_enemies(self, combatant):
        self_team = self.reverse_team_book[combatant]
        other_team = self.Color.RED if self_team is self.Color.BLUE else self.Color.BLUE
        try:
            return self.team_book[other_team]
        except KeyError:
            # This can only happen in test scenarios
            return []

File: c++/core/teams.cpp
#include "teams.hpp"

namespace enc
{

  std::unique_ptr<Teams> Teams::_instance = nullptr;

  Teams &Teams::getInstance()
  {
    if(!_instance)
      {
        _instance = std::unique_ptr<Teams>(new Teams());
      }
    return *_instance;
  }

  void Teams::resetInstance() { _instance.reset(new Teams()); }

  void Teams::addCombatantToTeam(Combatant *combatant, Color teamColor)
  {
    _idToCombatant[combatant->_instanceId] = combatant;
    _combatantIdToTeamColor[combatant->_instanceId] = teamColor;
    _colorToCombatantIds[teamColor].push_back(combatant->_instanceId);
    combatant->setTeamColor(teamColor);
  }

  void Teams::replaceCombatant(const Combatant &combatantOld, const Combatant &combatantNew)
  {
    Combatant *combatant = getCombatantById(combatantNew._instanceId);

    Color teamColor = _combatantIdToTeamColor[combatantOld._instanceId];
    _combatantIdToTeamColor[combatantNew._instanceId] = teamColor;
    _combatantIdToTeamColor.erase(combatantOld._instanceId);

    auto &teamIds = _colorToCombatantIds[teamColor];
    teamIds.erase(std::remove(teamIds.begin(), teamIds.end(), combatantOld._instanceId), teamIds.end());
    teamIds.push_back(combatantNew._instanceId);

    _idToCombatant.erase(combatantOld._instanceId);
    _idToCombatant[combatantNew._instanceId] = combatant;
  }

  std::string Teams::getTeamColorCode(const Combatant &combatant) const { return toString(_combatantIdToTeamColor.at(combatant._instanceId)); }

  Color Teams::getTeamColor(const Combatant &combatant) const { return _combatantIdToTeamColor.at(combatant._instanceId); }

  std::vector<Color> Teams::getSurvivingTeams() const
  {
    std::vector<Color> survivingTeams;
    for(const auto &[color, combatantIds] : _colorToCombatantIds)
      {
        if(std::any_of(combatantIds.begin(), combatantIds.end(), [this](int id) { return _idToCombatant.at(id)->isAlive(); }))
          {
            survivingTeams.push_back(color);
          }
      }
    return survivingTeams;
  }

  std::pair<int, int> Teams::getDeathCount() const
  {
    auto getDeathCountForTeam = [this](Color color) {
      const auto &combatantIds = _colorToCombatantIds.at(color);
      return combatantIds.size()
             - std::count_if(combatantIds.begin(), combatantIds.end(), [this](int id) { return _idToCombatant.at(id)->isAlive(); });
    };
    return {getDeathCountForTeam(Color::BLUE), getDeathCountForTeam(Color::RED)};
  }

  std::vector<Color> Teams::getTeamColors() const
  {
    std::vector<Color> colors;
    for(const auto &[color, _] : _colorToCombatantIds)
      {
        colors.push_back(color);
      }
    return colors;
  }

  bool Teams::areAllies(const Combatant &first, const Combatant &second) const
  {
    return _combatantIdToTeamColor.at(first._instanceId) == _combatantIdToTeamColor.at(second._instanceId);
  }

  bool Teams::areEnemies(const Combatant &first, const Combatant &second) const
  {
    return _combatantIdToTeamColor.at(first._instanceId) != _combatantIdToTeamColor.at(second._instanceId);
  }

  Color Teams::getTeam(const Combatant &combatant) const { return _combatantIdToTeamColor.at(combatant._instanceId); }

  std::vector<Combatant *> Teams::getAllAllies(const Combatant &combatant) const
  {
    std::vector<int> allyIds = _colorToCombatantIds.at(_combatantIdToTeamColor.at(combatant._instanceId));
    std::vector<Combatant *> allies;
    allies.reserve(allyIds.size());
    for(auto id : allyIds)
      {
        allies.push_back(_idToCombatant.at(id));
      }
    return allies;
  }

  std::vector<Combatant *> Teams::getAllEnemies(const Combatant &combatant) const
  {
    Color otherTeam = (_combatantIdToTeamColor.at(combatant._instanceId) == Color::BLUE) ? Color::RED : Color::BLUE;
    std::vector<int> enemyIds = _colorToCombatantIds.at(otherTeam);
    std::vector<Combatant *> enemies;
    enemies.reserve(enemyIds.size());
    for(auto id : enemyIds)
      {
        enemies.push_back(_idToCombatant.at(id));
      }
    return enemies;
  }

  std::vector<Combatant *> Teams::getAliveEnemies(const Combatant &combatant) const
  {
    std::vector<Combatant *> result;
    Color combatantTeam = getTeam(combatant);
    for(const auto &[color, ids] : _colorToCombatantIds)
      {
        if(color != combatantTeam)
          {
            for(int id : ids)
              {
                if(auto enemy = _idToCombatant.at(id); enemy && enemy->isAlive())
                  {
                    result.push_back(enemy);
                  }
              }
          }
      }
    return result;
  }

  std::vector<Combatant *> Teams::getAliveNonSwallowedEnemies(const Combatant &combatant) const
  {
    std::vector<Combatant *> result;
    Color combatantTeam = getTeam(combatant);
    for(const auto &[color, ids] : _colorToCombatantIds)
      {
        if(color != combatantTeam)
          {
            for(int id : ids)
              {
                if(auto enemy = _idToCombatant.at(id); enemy && enemy->isAlive() && !enemy->getSwallower())
                  {
                    result.push_back(enemy);
                  }
              }
          }
      }
    return result;
  }

  std::vector<Combatant *> Teams::getAliveNonSwallowedAllies(const Combatant &combatant) const
  {
    std::vector<Combatant *> result;
    Color combatantTeam = getTeam(combatant);
    for(int id : _colorToCombatantIds.at(combatantTeam))
      {
        if(Combatant *ally = _idToCombatant.at(id); ally && ally->isAlive() && !ally->getSwallower() && *ally != combatant)
          {
            result.push_back(ally);
          }
      }
    return result;
  }

  std::vector<Combatant *> Teams::getAliveCombatants(const Combatant &excludeCombatant) const
  {
    std::vector<Combatant *> result;
    for(const auto &[id, combatant] : _idToCombatant)
      {
        if(Combatant *cmbt = combatant; cmbt && cmbt->isAlive() && *cmbt != excludeCombatant)
          {
            result.push_back(combatant);
          }
      }
    return result;
  }

  std::vector<Combatant *> Teams::getAliveAllies(const Combatant &combatant) const
  {
    std::vector<Combatant *> result;
    Color combatantTeam = getTeam(combatant);
    for(int id : _colorToCombatantIds.at(combatantTeam))
      {
        if(Combatant *ally = _idToCombatant.at(id); ally && ally->isAlive() && *ally != combatant)
          {
            result.push_back(ally);
          }
      }
    return result;
  }

  Combatant *Teams::getCombatantById(int id) const { return _idToCombatant.at(id); }

} // namespace enc
